components/selection/selection-header.js

```diff
@@ -0,0 +1,186 @@
+ * @slot - Responsive container using `d2l-overflow-group` for `d2l-selection-action` elements
+ */
+export class SelectionHeader extends SelectionObserverMixin(RtlMixin(LocalizeCoreElement(LitElement))) {
+
+	static get properties() {
+		return {
+			/**
+			 * Whether to render select-all and selection summary
+			 * @type {boolean}
+			 */
+			noSelection: { type: Boolean, attribute: 'no-selection' },
+			/**
+			 * Disables sticky positioning for the header
+			 * @type {boolean}
+			 */
+			noSticky: { type: Boolean, attribute: 'no-sticky' },
+			/**
+			 * Whether all pages can be selected
+			 * @type {boolean}
+			 */
+			selectAllPagesAllowed: { type: Boolean, attribute: 'select-all-pages-allowed' },
+			_hasActions: { state: true },
+			_scrolled: { type: Boolean, reflect: true }
+		};
+	}
+
+	static get styles() {
+		return css`
+			:host {
+				background-color: var(--d2l-selection-header-background-color, white);
+				display: block;
+				position: sticky;
+				top: 0;
+			}
+			:host([no-sticky]) {
+				background-color: transparent;
+				position: static;
+			}
+			.d2l-selection-header-shadow {
+				transition: box-shadow 200ms ease-out;
+			}
+			:host([_scrolled]) .d2l-selection-header-shadow {
+				background-color: var(--d2l-selection-header-background-color, white);
+				bottom: -4px;
+				box-shadow: 0 8px 12px -9px rgba(0, 0, 0, 0.3);
+				clip: rect(30px, auto, 200px, auto);
+				height: 40px;
+				position: absolute;
+				width: 100%;
+				z-index: -1;
+			}
+			:host([hidden]) {
+				display: none;
+			}
+			.d2l-selection-header-container {
+				align-items: center;
+				display: flex;
+				margin-bottom: 6px;
+				margin-top: 6px;
+				min-height: 54px;
+			}
+			.d2l-selection-header-container-slim {
+				min-height: 36px;
+			}
+			d2l-selection-select-all {
+				flex: none;
+			}
+			d2l-selection-summary {
+				flex: none;
+				margin-left: 0.9rem;
+			}
+			:host([dir="rtl"]) d2l-selection-summary {
+				margin-left: 0;
+				margin-right: 0.9rem;
+			}
+			d2l-selection-select-all-pages {
+				flex: none;
+				margin-left: 0.45rem;
+			}
+			:host([dir="rtl"]) d2l-selection-select-all-pages {
+				margin-left: 0;
+				margin-right: 0.45rem;
+			}
+			.d2l-selection-header-actions {
+				--d2l-overflow-group-justify-content: flex-end;
+				flex: auto;
+				text-align: right;
+			}
+			:host([dir="rtl"]) .d2l-selection-header-actions {
+				text-align: left;
+			}
+			.d2l-sticky-edge {
+				margin-top: -1px;
+				position: absolute;
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this.noSelection = false;
+		this.noSticky = false;
+		this.selectAllPagesAllowed = false;
+		this._scrolled = false;
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		this._stickyObserverDisconnect();
+	}
+
+	render() {
+		const classes = this._getSelectionHeaderContainerClasses();
+		return html`
+			<div class="d2l-sticky-edge"></div>
```

Comment:
This replaces the "d2l-list-top-sentinel" that was previously being rendered in `list.js`, *outside* of the list-header. I've moved it inside the header but with `position: absolute` and `margin-top: -1px` to achieve the same effect - when it scrolls out of view, the rest of the header is treated as stuck (`._scrolled`).

Reply:
Previously I think the sentinel element was just a zero-height `div` so layout wasn't an issue. Was there a problem you encountered that necessitates the positioning/negative margin styles?

Reply:
Yep, it's still zero-height.

Previously the sentinel was outside of the header, so it would scroll off the page before the header got stuck. My goal here was to keep the sentinel inside the header, to keep everything encapsulated (rather than having to define the sentinel separately in e.g. `list.js`).

So to accomplish that, I:

- put it inside the (sticky) header
- gave it an absolute position and negative margin to render it 1px above the top edge of the sticky header
    - this way it goes offscreen when the header gets stuck, same as the external sentinel did
    - absolute position also means it doesn't affect the flow of the other elements
- note that `absolute` works a bit differently in terms of horizontal size, so I had to give it `left: 0` and `right: 0` CSS to get it to span the width of the container
    - otherwise it's triggered as fully offscreen when scrolling the left edge of the header offscreen
    - `width: 100%` would also have worked
    - positioning is all relative to the selection-header (nearest [positioned ancestor](https://developer.mozilla.org/en-US/docs/Web/CSS/position#types_of_positioning)) so `left`/`right` should be safe to use

This approach should also work for making things sticky against left/right/bottom as well.

Reply:
Sounds great. Thanks for the explanation! 😄

Reply:
It just occurred to me that I could do `top: -1px` instead of `margin-top`, should be equivalent.